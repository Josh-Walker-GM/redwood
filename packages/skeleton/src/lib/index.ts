import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
