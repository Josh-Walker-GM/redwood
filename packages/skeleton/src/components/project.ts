import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
