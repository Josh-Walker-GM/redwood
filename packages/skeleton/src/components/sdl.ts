import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'
import {
  ExportNamedDeclaration,
  isIdentifier,
  isTemplateElement,
  isVariableDeclaration,
  isVariableDeclarator,
  VariableDeclarator,
} from '@babel/types'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import { getASTFromFile } from '../lib/ast'

import { RedwoodSkeleton } from './base'
import type { RedwoodProject } from './project'

export class RedwoodSDL extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly gql: string | undefined

  constructor(filepath: string) {
    super(filepath)

    const ast = getASTFromFile(this.filepath)
    const namedExports: ExportNamedDeclaration[] = []
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        namedExports.push(path.node)
      },
    })

    let schemaExportVariableDeclarator: VariableDeclarator | undefined
    namedExports.forEach((namedExport) => {
      if (
        namedExport.declaration != null &&
        isVariableDeclaration(namedExport.declaration) &&
        isVariableDeclarator(namedExport.declaration.declarations[0]) &&
        isIdentifier(namedExport.declaration.declarations[0].id) &&
        namedExport.declaration.declarations[0].id.name === 'schema'
      ) {
        schemaExportVariableDeclarator = namedExport.declaration.declarations[0]
      }
    })
    if (schemaExportVariableDeclarator === undefined) {
      this.errors.push("No 'schema' export could be found")
    } else {
      switch (schemaExportVariableDeclarator.init?.type) {
        case 'TaggedTemplateExpression':
          if (
            schemaExportVariableDeclarator.init.quasi.quasis[0] !== undefined &&
            isTemplateElement(
              schemaExportVariableDeclarator.init.quasi.quasis[0]
            )
          ) {
            this.gql =
              schemaExportVariableDeclarator.init.quasi.quasis[0].value.cooked
          }
          break

        default:
          this.warnings.push(
            `Unable to extract the gql (${schemaExportVariableDeclarator.init?.type})`
          )
          break
      }
    }
    if (this.gql === undefined) {
      this.warnings.push('Could not extract the gql')
    }

    // TODO: Process the gql
  }
}

export function extractSDL(filepath: string): RedwoodSDL {
  return new RedwoodSDL(filepath)
}

export function extractSDLs(
  project: RedwoodProject | undefined = undefined
): RedwoodSDL[] {
  const sdls: RedwoodSDL[] = []

  const sdlPath = project
    ? getPaths(project.filepath).api.graphql
    : getPaths().api.graphql

  // TODO: Confirm this the condition to detect a sdl
  // SDLs must be defined within the graphql directory (at the top level?)
  const getSDLFiles = (directory: string) => {
    const SDLFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isFile()) {
        SDLFiles.push(path.join(directory, content))
      }
    })
    return SDLFiles
  }

  const sdlFiles = getSDLFiles(sdlPath)
  sdlFiles.forEach((sdlPath) => {
    if (!sdlPath.endsWith('.keep')) {
      sdls.push(extractSDL(sdlPath))
    }
  })

  return sdls
}
