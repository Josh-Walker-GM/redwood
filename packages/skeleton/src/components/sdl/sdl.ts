import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'
import {
  ExportNamedDeclaration,
  isIdentifier,
  isTemplateElement,
  isVariableDeclaration,
  isVariableDeclarator,
  VariableDeclarator,
} from '@babel/types'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import { getASTFromFile } from '../../lib/ast'
import type { RedwoodProject } from '../project'
import { RedwoodSkeleton } from '../skeleton'

import { extractMutations, RedwoodSDLMutation } from './mutations'
import { extractQueries, RedwoodSDLQuery } from './query'

export class RedwoodSDL extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  // TODO: Maybe we don't need the full gql
  readonly gql: string | undefined

  readonly queries: RedwoodSDLQuery[] | undefined
  readonly mutations: RedwoodSDLMutation[] | undefined

  constructor(filepath: string) {
    super(filepath)

    const ast = getASTFromFile(this.filepath)
    const namedExports: ExportNamedDeclaration[] = []
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        namedExports.push(path.node)
      },
    })

    let schemaExportVariableDeclarator: VariableDeclarator | undefined
    namedExports.forEach((namedExport) => {
      if (
        namedExport.declaration != null &&
        isVariableDeclaration(namedExport.declaration) &&
        isVariableDeclarator(namedExport.declaration.declarations[0]) &&
        isIdentifier(namedExport.declaration.declarations[0].id) &&
        namedExport.declaration.declarations[0].id.name === 'schema'
      ) {
        schemaExportVariableDeclarator = namedExport.declaration.declarations[0]
      }
    })
    if (schemaExportVariableDeclarator === undefined) {
      this.errors.push("No 'schema' export could be found")
    } else {
      switch (schemaExportVariableDeclarator.init?.type) {
        case 'TaggedTemplateExpression':
          if (
            schemaExportVariableDeclarator.init.quasi.quasis[0] !== undefined &&
            isTemplateElement(
              schemaExportVariableDeclarator.init.quasi.quasis[0]
            )
          ) {
            this.gql =
              schemaExportVariableDeclarator.init.quasi.quasis[0].value.cooked
          }
          break

        default:
          this.warnings.push(
            `Unable to extract the gql (${schemaExportVariableDeclarator.init?.type})`
          )
          break
      }
    }
    if (this.gql === undefined) {
      this.errors.push('Could not extract the gql')
    }

    this.queries = extractQueries(this)
    this.mutations = extractMutations(this)
  }
}

export function extractSDL(filepath: string): RedwoodSDL {
  return new RedwoodSDL(filepath)
}

export function extractSDLs(
  project: RedwoodProject | undefined = undefined
): RedwoodSDL[] {
  const sdls: RedwoodSDL[] = []

  const sdlPath = project
    ? getPaths(project.filepath).api.graphql
    : getPaths().api.graphql

  if (!fs.existsSync(sdlPath)) {
    return sdls
  }

  // TODO: Confirm this the condition to detect a sdl
  // SDLs must be defined within the graphql directory (at the top level?)
  const getSDLFiles = (directory: string) => {
    const SDLFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isFile()) {
        SDLFiles.push(path.join(directory, content))
      }
    })
    return SDLFiles
  }

  const sdlFiles = getSDLFiles(sdlPath)
  sdlFiles.forEach((sdlPath) => {
    if (!sdlPath.endsWith('.keep')) {
      sdls.push(extractSDL(sdlPath))
    }
  })

  return sdls
}
