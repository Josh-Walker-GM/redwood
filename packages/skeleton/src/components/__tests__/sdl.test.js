import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodProject } from '../project'
import { extractSDLs } from '../sdl/sdl'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  it('returns the correct SDLs without a project', () => {
    const sdls = extractSDLs(undefined)
    sdls.forEach((sdl) => {
      sdl.filepath = stripAndFormatPathForTesting(sdl.filepath, PROJECT_PATH)
      sdl.queries.forEach((query) => {
        query.filepath = stripAndFormatPathForTesting(
          query.filepath,
          PROJECT_PATH
        )
      })
      sdl.mutations.forEach((mutation) => {
        mutation.filepath = stripAndFormatPathForTesting(
          mutation.filepath,
          PROJECT_PATH
        )
      })
      expect(sdl).toMatchSnapshot(sdl.filepath)
    })
    expect(sdls.length).toMatchSnapshot('SDL count')
  })

  it('returns the correct SDLs with a project', () => {
    const project = RedwoodProject.getProject({
      pathWithinProject: PROJECT_PATH,
      readFromCache: false,
      insertIntoCache: false,
    })
    const sdls = extractSDLs(project)
    sdls.forEach((sdl) => {
      sdl.filepath = stripAndFormatPathForTesting(sdl.filepath, PROJECT_PATH)
      sdl.queries.forEach((query) => {
        query.filepath = stripAndFormatPathForTesting(
          query.filepath,
          PROJECT_PATH
        )
      })
      sdl.mutations.forEach((mutation) => {
        mutation.filepath = stripAndFormatPathForTesting(
          mutation.filepath,
          PROJECT_PATH
        )
      })
      expect(sdl).toMatchSnapshot(sdl.filepath)
    })
    expect(sdls.length).toMatchSnapshot('SDL count')
  })
})
