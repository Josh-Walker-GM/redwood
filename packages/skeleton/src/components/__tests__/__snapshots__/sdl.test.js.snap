// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct SDLs with a project: SDL count 1`] = `0`;

exports[`From within the empty-project fixture returns the correct SDLs without a project: SDL count 1`] = `0`;

exports[`From within the example-todo-main fixture returns the correct SDLs with a project: /api/src/graphql/currentUser.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/currentUser.sdl.ts",
  "gql": "
  type Query {
    currentUser: JSON
  }
",
  "name": "currentUser.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct SDLs with a project: /api/src/graphql/todos.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todos.sdl.js",
  "gql": "
  type Todo {
    id: Int!
    body: String!
    status: String!
  }

  type Query {
    todos: [Todo] @skipAuth
    todosCount: Int! @skipAuth
  }

  type Mutation {
    createTodo(body: String!): Todo @skipAuth
    updateTodoStatus(id: Int!, status: String!): Todo @skipAuth
    renameTodo(id: Int!, body: String!): Todo @skipAuth
  }
",
  "name": "todos.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct SDLs with a project: SDL count 1`] = `2`;

exports[`From within the example-todo-main fixture returns the correct SDLs without a project: /api/src/graphql/currentUser.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/currentUser.sdl.ts",
  "gql": "
  type Query {
    currentUser: JSON
  }
",
  "name": "currentUser.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct SDLs without a project: /api/src/graphql/todos.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todos.sdl.js",
  "gql": "
  type Todo {
    id: Int!
    body: String!
    status: String!
  }

  type Query {
    todos: [Todo] @skipAuth
    todosCount: Int! @skipAuth
  }

  type Mutation {
    createTodo(body: String!): Todo @skipAuth
    updateTodoStatus(id: Int!, status: String!): Todo @skipAuth
    renameTodo(id: Int!, body: String!): Todo @skipAuth
  }
",
  "name": "todos.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct SDLs without a project: SDL count 1`] = `2`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todos.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todos.sdl.js",
  "gql": "
  type Todo {
    id: Int!
    body: String!
    status: String!
  }

  type Query {
    todos: [Todo]
  }

  type Mutation {
    createTodo(body: String!): Todo
    updateTodoStatus(id: Int!, status: String!): Todo
    renameTodo(id: Int!, body: String!): Todo
  }
",
  "name": "todos.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosMutations.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosMutations.sdl.js",
  "gql": "
  type Query {
    deleteTodo(id: Int!): Todo! @requireAuth(roles: ["ADMIN"])
  }
",
  "name": "todosMutations.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithAuthInvalidRolesErrors.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthInvalidRolesErrors.sdl.js",
  "gql": "
  type Query {
    todosWithInvalidRoles: [Todo] @requireAuth(roles: ["admin", 12])
  }
",
  "name": "todosWithAuthInvalidRolesErrors.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithAuthMissingRoleError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthMissingRoleError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRoles: [Todo] @requireAuth(roles: [null, 12])
  }
",
  "name": "todosWithAuthMissingRoleError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithAuthRoles.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthRoles.sdl.js",
  "gql": "
  type Query {
    todosWithStringRole: [Todo] @requireAuth(roles: "admin")
    todosWithMultipleRoles: [Todo] @requireAuth(roles: ["admin", "editor"])
  }
",
  "name": "todosWithAuthRoles.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithBuiltInDirectives.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithBuiltInDirectives.sdl.js",
  "gql": "
  type Query {
    todosWithBuiltInDeprecatedDirective(id: Int!): Todo! @deprecated(reason: "Don't use this query anymore")
  }
",
  "name": "todosWithBuiltInDirectives.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithMissingAuthRolesAttributeError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithMissingAuthRolesAttributeError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRolesAttribute: [Todo] @requireAuth("ADMIN")
  }
",
  "name": "todosWithMissingAuthRolesAttributeError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithMissingAuthRolesAttributeNumericError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithMissingAuthRolesAttributeNumericError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRolesAttributeNumeric: [Todo] @requireAuth(42)
  }
",
  "name": "todosWithMissingAuthRolesAttributeNumericError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: /api/src/graphql/todosWithNumericRoleAuthError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithNumericRoleAuthError.sdl.js",
  "gql": "
  type Query {
    todosWithInvalidRole: [Todo] @requireAuth(roles: 12)
  }
",
  "name": "todosWithNumericRoleAuthError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs with a project: SDL count 1`] = `9`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todos.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todos.sdl.js",
  "gql": "
  type Todo {
    id: Int!
    body: String!
    status: String!
  }

  type Query {
    todos: [Todo]
  }

  type Mutation {
    createTodo(body: String!): Todo
    updateTodoStatus(id: Int!, status: String!): Todo
    renameTodo(id: Int!, body: String!): Todo
  }
",
  "name": "todos.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosMutations.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosMutations.sdl.js",
  "gql": "
  type Query {
    deleteTodo(id: Int!): Todo! @requireAuth(roles: ["ADMIN"])
  }
",
  "name": "todosMutations.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithAuthInvalidRolesErrors.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthInvalidRolesErrors.sdl.js",
  "gql": "
  type Query {
    todosWithInvalidRoles: [Todo] @requireAuth(roles: ["admin", 12])
  }
",
  "name": "todosWithAuthInvalidRolesErrors.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithAuthMissingRoleError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthMissingRoleError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRoles: [Todo] @requireAuth(roles: [null, 12])
  }
",
  "name": "todosWithAuthMissingRoleError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithAuthRoles.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithAuthRoles.sdl.js",
  "gql": "
  type Query {
    todosWithStringRole: [Todo] @requireAuth(roles: "admin")
    todosWithMultipleRoles: [Todo] @requireAuth(roles: ["admin", "editor"])
  }
",
  "name": "todosWithAuthRoles.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithBuiltInDirectives.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithBuiltInDirectives.sdl.js",
  "gql": "
  type Query {
    todosWithBuiltInDeprecatedDirective(id: Int!): Todo! @deprecated(reason: "Don't use this query anymore")
  }
",
  "name": "todosWithBuiltInDirectives.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithMissingAuthRolesAttributeError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithMissingAuthRolesAttributeError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRolesAttribute: [Todo] @requireAuth("ADMIN")
  }
",
  "name": "todosWithMissingAuthRolesAttributeError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithMissingAuthRolesAttributeNumericError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithMissingAuthRolesAttributeNumericError.sdl.js",
  "gql": "
  type Query {
    todosWithMissingRolesAttributeNumeric: [Todo] @requireAuth(42)
  }
",
  "name": "todosWithMissingAuthRolesAttributeNumericError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: /api/src/graphql/todosWithNumericRoleAuthError.sdl.js 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/todosWithNumericRoleAuthError.sdl.js",
  "gql": "
  type Query {
    todosWithInvalidRole: [Todo] @requireAuth(roles: 12)
  }
",
  "name": "todosWithNumericRoleAuthError.sdl",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct SDLs without a project: SDL count 1`] = `9`;

exports[`From within the test-project fixture returns the correct SDLs with a project: /api/src/graphql/contacts.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/contacts.sdl.ts",
  "gql": "
  type Contact {
    id: Int!
    name: String!
    email: String!
    message: String!
    createdAt: DateTime!
  }

  type Query {
    contacts: [Contact!]! @requireAuth
    contact(id: Int!): Contact @requireAuth
  }

  input CreateContactInput {
    name: String!
    email: String!
    message: String!
  }

  input UpdateContactInput {
    name: String
    email: String
    message: String
  }

  type Mutation {
    createContact(input: CreateContactInput!): Contact @skipAuth
    updateContact(id: Int!, input: UpdateContactInput!): Contact! @requireAuth
    deleteContact(id: Int!): Contact! @requireAuth(roles: ["ADMIN"])
  }
",
  "name": "contacts.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs with a project: /api/src/graphql/posts.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/posts.sdl.ts",
  "gql": "
  type Post {
    id: Int!
    title: String!
    body: String!
    authorId: Int!
    author: User!
    createdAt: DateTime!
  }

  type Query {
    posts: [Post!]! @skipAuth
    post(id: Int!): Post @skipAuth
  }

  input CreatePostInput {
    title: String!
    body: String!
    authorId: Int!
  }

  input UpdatePostInput {
    title: String
    body: String
    authorId: Int
  }

  type Mutation {
    createPost(input: CreatePostInput!): Post! @requireAuth
    updatePost(id: Int!, input: UpdatePostInput!): Post! @requireAuth
    deletePost(id: Int!): Post! @requireAuth
  }
",
  "name": "posts.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs with a project: /api/src/graphql/users.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/users.sdl.ts",
  "gql": "
  type User {
    id: Int!
    email: String!
    fullName: String!
    roles: String
    posts: [Post]!
  }

  type Query {
    user(id: Int!): User @skipAuth
  }

  input CreateUserInput {
    email: String!
    fullName: String!
    roles: String
  }

  input UpdateUserInput {
    email: String
    fullName: String
    roles: String
  }
",
  "name": "users.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs with a project: SDL count 1`] = `3`;

exports[`From within the test-project fixture returns the correct SDLs without a project: /api/src/graphql/contacts.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/contacts.sdl.ts",
  "gql": "
  type Contact {
    id: Int!
    name: String!
    email: String!
    message: String!
    createdAt: DateTime!
  }

  type Query {
    contacts: [Contact!]! @requireAuth
    contact(id: Int!): Contact @requireAuth
  }

  input CreateContactInput {
    name: String!
    email: String!
    message: String!
  }

  input UpdateContactInput {
    name: String
    email: String
    message: String
  }

  type Mutation {
    createContact(input: CreateContactInput!): Contact @skipAuth
    updateContact(id: Int!, input: UpdateContactInput!): Contact! @requireAuth
    deleteContact(id: Int!): Contact! @requireAuth(roles: ["ADMIN"])
  }
",
  "name": "contacts.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs without a project: /api/src/graphql/posts.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/posts.sdl.ts",
  "gql": "
  type Post {
    id: Int!
    title: String!
    body: String!
    authorId: Int!
    author: User!
    createdAt: DateTime!
  }

  type Query {
    posts: [Post!]! @skipAuth
    post(id: Int!): Post @skipAuth
  }

  input CreatePostInput {
    title: String!
    body: String!
    authorId: Int!
  }

  input UpdatePostInput {
    title: String
    body: String
    authorId: Int
  }

  type Mutation {
    createPost(input: CreatePostInput!): Post! @requireAuth
    updatePost(id: Int!, input: UpdatePostInput!): Post! @requireAuth
    deletePost(id: Int!): Post! @requireAuth
  }
",
  "name": "posts.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs without a project: /api/src/graphql/users.sdl.ts 1`] = `
RedwoodSDL {
  "errors": [],
  "filepath": "/api/src/graphql/users.sdl.ts",
  "gql": "
  type User {
    id: Int!
    email: String!
    fullName: String!
    roles: String
    posts: [Post]!
  }

  type Query {
    user(id: Int!): User @skipAuth
  }

  input CreateUserInput {
    email: String!
    fullName: String!
    roles: String
  }

  input UpdateUserInput {
    email: String
    fullName: String
    roles: String
  }
",
  "name": "users.sdl",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct SDLs without a project: SDL count 1`] = `3`;
